src/solc_0.7/claims/MultiGiveaway/MultiGiveawayWithERC20.sol

```diff
@@ -0,0 +1,97 @@
+//SPDX-License-Identifier: MIT
+pragma solidity 0.7.5;
+pragma experimental ABIEncoderV2;
+
+import "@openzeppelin/contracts/token/ERC721/IERC721Receiver.sol";
+import "@openzeppelin/contracts/token/ERC1155/IERC1155Receiver.sol";
+import "./ClaimERC1155ERC721ERC20.sol";
+import "../../common/BaseWithStorage/WithAdmin.sol";
+
+/// @title MultiGiveaway contract.
+/// @notice This contract manages claims for multiple token types.
+contract MultiGiveawayWithERC20 is WithAdmin, ClaimERC1155ERC721ERC20 {
+    bytes4 private constant ERC1155_RECEIVED = 0xf23a6e61;
+    bytes4 private constant ERC1155_BATCH_RECEIVED = 0xbc197c81;
+    bytes4 internal constant ERC721_RECEIVED = 0x150b7a02;
+    bytes4 internal constant ERC721_BATCH_RECEIVED = 0x4b808c46;
+
+    uint256 internal _counter = 1; // First giveaway is Multi_Giveaway_1
+
+    mapping(address => mapping(bytes32 => bool)) public claimed; // TODO: change to index mapping - see uniswap
+    mapping(bytes32 => uint256) internal _expiryTime;
+
+    constructor(address admin) {
+        _admin = admin;
+    }
+
+    /// @notice Function to set the merkle root hash for the claim data.
+    /// @param giveaway The giveaway number to change the merkle root hash for.
+    /// @param merkleRoot The merkle root hash of the claim data.
+    function setMerkleRoot(uint256 giveaway, bytes32 merkleRoot) external onlyAdmin {
+        _merkleRoots[giveaway] = merkleRoot;
+    }
+
+    /// @notice Function to set the merkle root hash for the claim data.
+    /// @param merkleRoot The merkle root hash of the claim data.
+    /// @param expiryTime The expiry time for the giveaway.
+    function addNewGiveaway(bytes32 merkleRoot, uint256 expiryTime) external onlyAdmin {
+        _merkleRoots[_counter] = merkleRoot;
```

Comment:
It might be easier to have one mapping with the merkle root hash being the key
the expiry could be the value itself